src/include/catalog/catalog.h

```diff
@@ -0,0 +1,325 @@
+ *
+ * pg_database, pg_tablespace and pg_settings are global catalogs
+ * (only a single instance). Other catalogs are "local" to each database.
+ * For example, a pg_class catalog exists for each database.
+ *
+ * OID assignment:
+ * In terrier, each table row is identified by an oid. Further, as strong
+ * typing is used in terrier, each (catalog) table has it's own oid type,
+ * e.g. db_oid_t, namespace_oid_t etc.
+ *
+ * The values for all oid types are generated from a single value space
+ * i.e. every oid regardless of type, is unique.
+ *
+ */
+class Catalog {
+ public:
+  /**
+   * Creates the (global) catalog object, and bootstraps.
+   * A default database is created. Its tables are created and populated.
+   *
+   * @param txn_manager the global transaction manager
+   * @param txn to be used for bootstrapping
+   */
+  Catalog(transaction::TransactionManager *txn_manager, transaction::TransactionContext *txn);
+
+  /**
+   * Create a database, and all its tables. Set initial content
+   * for the tables, and update the catalogs to reflect the new
+   * database (and its tables).
+   *
+   * @param txn transaction to use
+   * @param name of the database
+   */
+  void CreateDatabase(transaction::TransactionContext *txn, const std::string &name);
+
+  /**
+   * Delete a database.
+   *
+   * @param txn transaction to use
+   * @param db_name of the database
+   */
+  void DeleteDatabase(transaction::TransactionContext *txn, const std::string &db_name);
+
+  /**
+   * Create a table with schema
+   * @param txn transaction to use
+   * @param db_oid oid of the database
+   * @param table_name table name
+   * @param schema schema to use
+   */
+  table_oid_t CreateUserTable(transaction::TransactionContext *txn, db_oid_t db_oid, const std::string &table_name,
+                              const Schema &schema);
+
+  /**
+   * Delete a table
+   * @param txn transaction to use
+   * @param db_oid oid of the database
+   * @param table_name table to delete
+   */
+  void DeleteTable(transaction::TransactionContext *txn, db_oid_t db_oid, const std::string &table_name);
+
+  /**
+   * Return a database handle.
+   * @return the database handle
+   */
+  DatabaseHandle GetDatabaseHandle();
+
+  /**
+   * Return a tablespace handle.
+   * @return the tablespace handle
+   */
+  TablespaceHandle GetTablespaceHandle();
+
+  /**
+   * Return a settings handle.
+   * @return the settings handle
+   */
+  SettingsHandle GetSettingsHandle();
+
+  /**
+   * Get a pointer to the storage table.
+   * Supports both catalog tables, and user created tables.
+   *
+   * @param db_oid database that owns the table
+   * @param table_oid returns the storage table pointer for this table_oid
+   * @return a pointer to the catalog
+   * @throw out_of_range exception if either oid doesn't exist or the catalog doesn't exist.
+   */
+  std::shared_ptr<catalog::SqlTableRW> GetCatalogTable(db_oid_t db_oid, table_oid_t table_oid);
```

Comment:
Should this method be protected? Does anybody other than the catalog need it?

Reply:
Protected now.